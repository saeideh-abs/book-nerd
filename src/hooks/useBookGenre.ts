import { fetchGenresByBookIdQuery } from '@/services/genre'
import { useQuery } from '@tanstack/react-query'

export function useBookGenre(bookId: string) {
  return useQuery({
    queryKey: ['bookGenre', bookId],
    queryFn: async () => {
      const { data, error } = await fetchGenresByBookIdQuery(bookId)

      if (error) throw error

      return data
    },
    enabled: !!bookId,
  })
}
