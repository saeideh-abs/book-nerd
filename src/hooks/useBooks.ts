import { BookBoxItemType } from '@/types'
import { getFromAndTo, mergeAuthorAndRole } from '@/utils'
import { supabaseClient } from '@/utils/supabaseClient'
import { QueryData } from '@supabase/supabase-js'
import { useInfiniteQuery, useQuery } from '@tanstack/react-query'

export const booksQuery = supabaseClient
  .from('book')
  .select(`*, author (id, name), book_author (author_id, role)`)

export type BooksQueryType = QueryData<typeof booksQuery>

export function useBooks() {
  return useInfiniteQuery({
    queryKey: ['books'],
    queryFn: async ({ pageParam = 0 }) => {
      const [from, to] = getFromAndTo(pageParam)
      const { data, error } = await booksQuery.range(from, to)

      if (error) throw error

      return formatBooks(data)
    },
    select: data => ({
      pages: [...data.pages.flat()],
      pageParams: [...data.pageParams],
    }),
    getNextPageParam: (_, pages) => pages.length,
  })
}

export const formatBooks = (data: BooksQueryType | null): BookBoxItemType[] => {
  return data
    ? data.map(item => ({
        id: item.id,
        title: item.title ?? '',
        price: item.price,
        author: mergeAuthorAndRole(item.author, item.book_author),
        coverImg: item.cover_img,
      }))
    : []
}

// export const useBooks = () => {
//   const [booksData, setBooksData] = useState<BookBoxItemType[]>([])
//   const getData = useCallback(async () => {
//     const { data, error, status } = await booksQuery.range(0, 19)
//     if (error) throw error
//     setBooksData(formatBooks(data))
//     console.log(formatBooks(data), error, status)
//   }, [])
//   useEffect(() => {
//     getData()
//   }, [getData])
//   return [booksData]
// }

// return useQuery({
//   queryKey: ['books'],
//   queryFn: async () => {
//     const { data, error } = await booksQuery.range(20, 29)

//     if (error) throw error

//     return formatBooks(data)
//   },
// })
