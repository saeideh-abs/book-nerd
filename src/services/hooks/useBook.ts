import { BaseBookBoxItem } from '@/models/BaseBookBoxItem'
import { fetchBookByIdQuery } from '@/services/book'
import { useQuery } from '@tanstack/react-query'

export function useBook(bookId: string) {
  return useQuery({
    queryKey: ['book', bookId],
    queryFn: async () => {
      const { data, error } = await fetchBookByIdQuery(bookId)

      if (error) throw error

      return new BaseBookBoxItem(data)
    },
    enabled: !!bookId,
  })
}
