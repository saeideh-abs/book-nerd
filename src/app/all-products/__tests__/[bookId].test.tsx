import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import BookDetails from '../[bookId]/page'
import { useBook } from '@/services/hooks/useBook'
import { useBookGenre } from '@/services/hooks/useBookGenre'

declare global {
  interface String {
    toTitleCase(): string
  }
}

String.prototype.toTitleCase = function () {
  return this.toString()
}

// Mock the hooks
jest.mock('@/services/hooks/useBook')
jest.mock('@/services/hooks/useBookGenre')
jest.mock('next-intl', () => ({
  useTranslations: () => (key: string) => key,
}))

jest.mock('@supabase/auth-helpers-nextjs', () => ({
  createClientComponentClient: () => ({
    from: jest.fn().mockReturnThis(),
    select: jest.fn().mockReturnThis(),
    eq: jest.fn().mockReturnThis(),
    single: jest.fn().mockResolvedValue({ data: {}, error: null }),
    // Add any other Supabase client methods you use in your component
  }),
}))

describe('BookDetails', () => {
  const mockBookData = {
    id: '1',
    title: 'Test Book',
    author: [{ name: 'Test Author', main: true, role: null }],
    pages: 200,
    bookFormat: 'Paperback',
    description: 'Test description',
  }

  const mockGenreData = [
    { id: '1', name: 'fiction' },
    { id: '2', name: 'adventure' },
  ]

  beforeEach(() => {
    // Setup default mock returns
    ;(useBook as jest.Mock).mockReturnValue({
      data: mockBookData,
      error: null,
    })
    ;(useBookGenre as jest.Mock).mockReturnValue({
      data: mockGenreData,
      error: null,
    })
  })

  it('renders book details information correctly', () => {
    render(<BookDetails params={{ bookId: '1' }} />)
    expect(screen.getAllByText('Test Book')).toHaveLength(2)
    expect(screen.getAllByText('Test Author')).toHaveLength(2)
    expect(screen.getByText('200 pages, Paperback')).toBeInTheDocument()
    expect(screen.getByText('fiction')).toBeInTheDocument()
    expect(screen.getByText('adventure')).toBeInTheDocument()
    expect(screen.getByText('Test description')).toBeInTheDocument()
  })

  it('renders BookDetails page correctly', () => {
    const { container } = render(<BookDetails params={{ bookId: '1' }} />)
    expect(container).toMatchSnapshot()
  })
})
