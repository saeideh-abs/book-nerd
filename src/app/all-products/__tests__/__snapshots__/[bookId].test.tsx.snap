// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetails renders BookDetails page correctly 1`] = `
<div>
  <div
    class="flex flex-col md:flex-row gap-10 p-layoutXMd"
  >
    <div
      class="min-w-[260px]"
    >
      <div
        class="flex flex-col items-center p-6 border text-center border-gray-200 bg-white rounded-sm shadow-[0px_4px_10px_0px_rgba(0,0,0,0.15)] hover:opacity-50 h-auto w-fit md:w-full m-auto hover:cursor-auto"
      >
        <img
          alt=""
          class="w-[200px] h-auto"
          data-nimg="1"
          decoding="async"
          height="0"
          src=""
          style="color: transparent;"
          width="0"
        />
        <div
          class="flex flex-col justify-end grow"
        >
          <div
            class="text-[1.125rem] font-semibold leading-[1.75rem] mt-3 text-primary-500"
          >
            Test Book
          </div>
          <div
            class="text-[0.875rem] font-normal leading-[1.25rem] text-bnGray-400"
          >
            Test Author
          </div>
          <div
            class="text-[1rem] font-bold leading-[1.5rem] mt-3 text-secondary-500 invisible"
          >
            unknown
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="flex flex-col gap-3"
      >
        <div
          class="text-[1.875rem] font-semibold leading-[2.375rem] text-primary-500"
        >
          Test Book
        </div>
        <div
          class="text-[1.125rem] font-normal leading-[1.75rem] text-bnGray-400"
        >
          Test Author
        </div>
        <div
          class="text-[0.875rem] font-normal leading-[1.25rem] text-bnGray-400"
        >
          200 pages, 
          Paperback
        </div>
        <div
          class="flex flex-wrap gap-2"
        >
          <div
            class="px-2 py-1 rounded-md bg-secondary-50 text-secondary-600 text-xs font-medium"
          >
            fiction
          </div>
          <div
            class="px-2 py-1 rounded-md bg-secondary-50 text-secondary-600 text-xs font-medium"
          >
            adventure
          </div>
        </div>
        <div
          class="w-full h-[1px] bg-bnGray-500 my-3"
        />
        <div
          class="text-[1rem] font-normal leading-[1.5rem] text-primary-500"
        >
          Test description
        </div>
      </div>
    </div>
  </div>
</div>
`;
